url29ae058134_yuriy_chumak_jatha_tgz-pJ8-org_jatha_LispJ8/src/org/jatha/Lisp.java

```diff
@@ -0,0 +1,1243 @@
+          value = MACHINE.Execute(code, NIL);
+        } catch (Exception e2) {
+          System.out.println("Unable to evaluate " + input + "\n  " + e2);
+          continue;
+        }
+
+        // PRINT
+        value.prin1();
+      }
+    }
+  }
+
+
+  /**
+   * Returns the LISP compiler used by this instance of Jatha.
+   */
+  public LispCompiler getCompiler()
+  {
+    return COMPILER;
+  }
+
+
+  /**
+   * Returns the LISP Parser used by this instance of Jatha.
+   */
+  public LispParser getParser()
+  {
+    return PARSER;
+  }
+
+  /**
+   * Returns the Symbol Table used by this instance of Jatha.
+   */
+  public SymbolTable getSymbolTable()
+  {
+    return SYMTAB;
+  }
+
+  /**
+   * Parses a string and returns the first form in the string.
+   * <br>caseSensitivity:
+   * <ul>
+   *   <li>LispParser.UPCASE (the default)</li>
+   *   <li>LispParser.DOWNCASE</li>
+   *   <li>LispParser.PRESERVE</li>
+   * </ul>
+   */
+  public LispValue parse(String s, int caseSensitivity)
+    throws EOFException
+  {
+    return new LispParser(this, s, caseSensitivity).parse();
+  }
+
+
+  /**
+   * Parses a string and returns the first form in the string.
+   */
+  public LispValue parse(String s)
+    throws EOFException
+  {
+    return parse(s, LispParser.UPCASE);
+  }
+
+  /**
+   * Loads the contents of a Reader (stream).
+   * Useful for loading from a jar file.
+   * Contributed by Stephen Starkey.
+   */
+  public LispValue load(Reader in) throws CompilerException
+  {
+    return load(in, false);
+  }
+
+  /**
+   * Loads the contents of a Reader (stream).
+   * Useful for loading from a jar file.
+   * Contributed by Stephen Starkey.
+   */
+  public LispValue load(Reader in, boolean verbose)
+      throws CompilerException
+  {
+    // System.err.println("Loading: verbose is " + verbose);
+
+    BufferedReader buff = new BufferedReader(in);
+
+    LispParser fileparser = new LispParser(this, buff);
+    LispValue  input, code;
+    boolean    atLeastOneResult = false;
+
+    // Read and Eval stream until EOF.
+    try {
+      while (true)
+      {
+        input = fileparser.parse();
+
+        code  = COMPILER.compile(MACHINE, input, NIL);
+
+        LispValue value = MACHINE.Execute(code, NIL);
+        atLeastOneResult = true;
+
+        if (verbose)
+        {
+          System.out.println(value.toString());
+        }
+      }
+    } catch (IOException ioe) {
+      try {
+        in.close();
+      } catch (IOException e2) {
+        return T;
+      }
+      catch (Exception ex)
+      {
+    	final Exception ex2 = ex;
+    	throw new LispValueNotAConsException();
+      }
+    }
+
+    if (atLeastOneResult)
+        return T;
+      else
+        return NIL;
+  }
+
+
+  /** Loads a file.
+   * Argument is guaranteed to be a LispString.
+   */
+  public LispValue load(LispValue filenameVal)
+  {
+    String filename = ((LispString) filenameVal).getValue();
+
+    try {
+      return load(new FileReader(filename));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *FileReader*. Currently, there are execution paths that do not contain closure statements, for example, when Lisp.load throws an exception. Ensure the object returned by *FileReader()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).